### PR TITLE
[mysql] Support the rate limit when Mysql parallel read data output

### PR DESCRIPTION
**Pull request by fuyun2024**
issues : #510  
This PR  support the rate limit when Mysql parallel reads data output

### CODE REVIEW

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/config/MySqlSourceOptions.java [hunk 1]
```diff
@@ -127,6 +127,12 @@ public class MySqlSourceOptions {
                     .defaultValue(20)
                     .withDescription("The connection pool size.");
 
+    public static final ConfigOption<Integer> OUTPUT_RATE_LIMIT =
+            ConfigOptions.key("out.rate-limit")
+                    .intType()
+                    .defaultValue(0)
```
Comment (GOODBOY008) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/config/MySqlSourceOptions.java:133]:
```suggestion
                    .noDefaultValue()
```
I think this config option with no default value will be better.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 2]
```diff
@@ -101,6 +102,7 @@ public DynamicTableSource createDynamicTableSource(Context context) {
         Duration connectTimeout = config.get(CONNECT_TIMEOUT);
         int connectMaxRetries = config.get(CONNECT_MAX_RETRIES);
         int connectionPoolSize = config.get(CONNECTION_POOL_SIZE);
+        long outputRateLimit = config.get(OUTPUT_RATE_LIMIT);
```
Comment (GOODBOY008) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:105]: Maybe we can invork `OUTPUT_RATE_LIMIT` validate method here like `DATABASE_NAME` and `startupOptions` config.

**Review by GOODBOY008 — COMMENTED** [review]
@fuyun2024 Thanks for your contribution , I leave some comments.

**Comment by fuyun2024** [comment]
@GOODBOY008  I have fixed and resubmitted code

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/config/MySqlSourceOptions.java (hunk 1 above)
Comment (fuyun2024) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/config/MySqlSourceOptions.java]: I tried to modify it, but I found that there are some 'nullpointerException', so I think the default value is 0

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/MySqlSource.java [hunk 3]
```diff
@@ -227,4 +230,15 @@ public SimpleVersionedSerializer<PendingSplitsState> getEnumeratorCheckpointSeri
     public TypeInformation<T> getProducedType() {
         return deserializationSchema.getProducedType();
     }
+
+    private DebeziumDeserializationSchema<T> getActualDebeziumDeserializationSchema(
+            Long outputRateLimit, SourceReaderContext readerContext) {
+        if (outputRateLimit != null && outputRateLimit > 0) {
```
Comment (GOODBOY008) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/MySqlSource.java:236]:
```suggestion
            long outputRateLimit, SourceReaderContext readerContext) {
        if (outputRateLimit > 0) {
```

File: docs/content/connectors/mysql-cdc(ZH).md [hunk 4]
```diff
@@ -259,6 +259,13 @@ Flink SQL> SELECT * FROM orders;
           <td>Integer</td>
           <td>连接池大小。</td>
     </tr>
+    <tr>
+          <td>output.rate-limit</td>
+          <td>optional</td>
+          <td style="word-wrap: break-word;">0</td>
+          <td>Long</td>
+          <td>每秒输出记录速率限制。您可以设置此参数，以防止在快照收集期间影响数据库的正常使用。它还可以防止流量抖动，提高稳定性。默认 0 (无限制)。</td>
```
Comment (GOODBOY008) [inline docs/content/connectors/mysql-cdc(ZH).md:267]:
```suggestion
          <td>限制每秒输出记录数量。你可以通过设置此参数，以防止在快照期间影响数据库的正常使用。除此以外还可以防止流量抖动，提高稳定性。默认 0 (无限制)。</td>
```

File: flink-connector-debezium/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+    private enum OperatingMode {
+        READING,
+        OUTPUT_NOT_INITIALIZED,
+        SOURCE_STOPPED,
+        DATA_FINISHED
+    }
+
+    private InternalSourceReaderMetricGroup sourceMetricGroup;
+
+    private @Nullable LatencyMarkerEmitter<OUT> latencyMarerEmitter;
```
Comment (GOODBOY008) [inline flink-connector-debezium/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:155]: Typo `Marer`. You can copy lastest souce code from flink.

**Review by GOODBOY008 — COMMENTED** [review]
@fuyun2024 Thanks for your contribution , I have left litlle comment

**Comment by ChenShuai1981** [comment]
How about this PR now? Can you fix conflicts and merge into master branch? We need this feature

**Comment by yuxiqian** [comment]
Hi @fuyun2024, thanks for your great contribution! Also sorry for the delay of this PR. Could you please rebase it to latest `master` branch since there’s been lots of changes in Flink CDC repo since your original commit? Kindly reminder that `com.ververica.cdc.debezium` package has been moved to `org.apache.flink.cdc.debezium`.

**Pull request closed without merging**